### ISSUE TITLE
`tf.reverse_sequence()` doesn't work with arguments of totally unknown shape.

### ISSUE THREAD

**Issue by mrry** [tensorflow/tensorflow#1816]
From [Stack Overflow](http://stackoverflow.com/questions/36480456/dynamic-rnn-and-array-ops-reverse-sequence-problems):

> I am trying to reverse my inputs with array_ops.reverse_sequence() before sending it to dynamic_rnn(), the inference graph can be build with no problem, but when building the training graph, I got the following error:
> 
> 
> ```
>   Traceback (most recent call last):
>   File "bin/trainer.py", line 158, in <module>
>     kmer_len=args.kmer_len)
>   File "/home/ubuntu/GIT/IvyMike/ivymike/base_model.py", line 193, in run_training
>     train_op = model.training(loss, learning_rate)
>   File "/home/ubuntu/GIT/IvyMike/ivymike/base_model.py", line 100, in training
>     train_op = optimizer.minimize(loss, global_step=global_step)
>   File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/training/optimizer.py", line 190, in minimize
>     colocate_gradients_with_ops=colocate_gradients_with_ops)
>   File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/training/optimizer.py", line 241, in compute_gradients
>     colocate_gradients_with_ops=colocate_gradients_with_ops)
>   File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/gradients.py", line 481, in gradients
>     in_grads = _AsList(grad_fn(op, *out_grads))
>   File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/array_grad.py", line 307, in _ReverseSequenceGrad
>     seq_lengths=seq_lengths),
>   File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/gen_array_ops.py", line 1143, in reverse_sequence
>     batch_dim=batch_dim, name=name)
>   File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/op_def_library.py", line 655, in apply_op
>     op_def=op_def)
>   File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 2119, in create_op
>     set_shapes_for_outputs(ret)
>   File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 1586, in set_shapes_for_outputs
>     shapes = shape_func(op)
>   File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/array_ops.py", line 1257, in _ReverseSequenceShape
>     (batch_dim, input_shape.ndims))
> TypeError: %d format: a number is required, not NoneType
> ```

The problem appears to arise when `input_shape.ndims` is `None` (which is a valid possibility).

**Comment by zhangjiajie** [comment]
For some reasons I am not allowed to post my code here, but the bug occurs when I try to stack multiple layers of BLSTM using dynamic_rnn() (see this related request: https://github.com/tensorflow/tensorflow/issues/1779). For the backward pass, I first reversed the input using array_ops.reverse_sequence(), then reversed the output using the same function. I also used feed forward layers in between the BLSTMs. 

However, with one BLSTM layer, everything is fine.  

Thanks for all the help!
